### ISSUE TITLE
'Broken Pipe' Error When Pushing Image

### ISSUE THREAD

**Issue by helenxie-bit** [docker/for-mac#7349]
### Description

I have built an image and want to push it to Docker Hub, but I keep encountering the following error:
failed to copy: failed to do request: Put "https://registry-1.docker.io/v2/helenxiehz428/test_llm/blobs/uploads/9ec1e7d0-43cc-4c38-9315-cb193824e0e2?_state=dmk14uQV81DeUHjT3obo3x0hIJNtFTNYc_6pw5thzhV7Ik5hbWUiOiJoZWxlbnhpZWh6NDI4L3Rlc3RfbGxtIiwiVVVJRCI6IjllYzFlN2QwLTQzY2MtNGMzOC05MzE1LWNiMTkzODI0ZTBlMiIsIk9mZnNldCI6MCwiU3RhcnRlZEF0IjoiMjAyNC0wNy0xMFQwNToxODoyNC45MjIwMTgxNDFaIn0%3D&digest=sha256%3A0a6fb9455297c44eac98ceb6ee7c2621b21748cf24cc571ac49cd73c38945eb8": write tcp 192.168.65.3:60764->192.168.65.1:3128: write: broken pipe

### Reproduce

# Build the Docker image and tag it with a name
docker build -t <image_name> .

# Tag the Docker image with Docker Hub username
docker tag <image_name> <dockerhub_username>/<image_name>

# Push the tagged Docker image to Docker Hub
docker push <dockerhub_username>/<image_name>

### Expected behavior

Successfully push the image to the Docker Hub

### docker version

```bash
Client: Docker Engine - Community
 Version:           26.1.3
 API version:       1.45
 Go version:        go1.22.3
 Git commit:        b72abbb6f0
 Built:             Thu May 16 07:47:24 2024
 OS/Arch:           darwin/arm64
 Context:           desktop-linux

Server: Docker Desktop 4.31.0 (153195)
 Engine:
  Version:          26.1.4
  API version:      1.45 (minimum version 1.24)
  Go version:       go1.21.11
  Git commit:       de5c9cf
  Built:            Wed Jun  5 11:29:12 2024
  OS/Arch:          linux/arm64
  Experimental:     false
 containerd:
  Version:          1.6.33
  GitCommit:        d2d58213f83a351ca8f528a95fbd145f5654e957
 runc:
  Version:          1.1.12
  GitCommit:        v1.1.12-0-g51d5e94
 docker-init:
  Version:          0.19.0
  GitCommit:        de40ad0
```

### docker info

```bash
Client: Docker Engine - Community
 Version:    26.1.3
 Context:    desktop-linux
 Debug Mode: false
 Plugins:
  buildx: Docker Buildx (Docker Inc.)
    Version:  v0.14.1-desktop.1
    Path:     /Users/helen/.docker/cli-plugins/docker-buildx
  compose: Docker Compose (Docker Inc.)
    Version:  v2.27.1-desktop.1
    Path:     /Users/helen/.docker/cli-plugins/docker-compose
  debug: Get a shell into any image or container (Docker Inc.)
    Version:  0.0.32
    Path:     /Users/helen/.docker/cli-plugins/docker-debug
  dev: Docker Dev Environments (Docker Inc.)
    Version:  v0.1.2
    Path:     /Users/helen/.docker/cli-plugins/docker-dev
  extension: Manages Docker extensions (Docker Inc.)
    Version:  v0.2.24
    Path:     /Users/helen/.docker/cli-plugins/docker-extension
  feedback: Provide feedback, right in your terminal! (Docker Inc.)
    Version:  v1.0.5
    Path:     /Users/helen/.docker/cli-plugins/docker-feedback
  init: Creates Docker-related starter files for your project (Docker Inc.)
    Version:  v1.2.0
    Path:     /Users/helen/.docker/cli-plugins/docker-init
  sbom: View the packaged-based Software Bill Of Materials (SBOM) for an image (Anchore Inc.)
    Version:  0.6.0
    Path:     /Users/helen/.docker/cli-plugins/docker-sbom
  scout: Docker Scout (Docker Inc.)
    Version:  v1.9.3
    Path:     /Users/helen/.docker/cli-plugins/docker-scout

Server:
 Containers: 1
  Running: 1
  Paused: 0
  Stopped: 0
 Images: 5
 Server Version: 26.1.4
 Storage Driver: overlayfs
  driver-type: io.containerd.snapshotter.v1
 Logging Driver: json-file
 Cgroup Driver: cgroupfs
 Cgroup Version: 2
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local splunk syslog
 Swarm: inactive
 Runtimes: io.containerd.runc.v2 runc
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: d2d58213f83a351ca8f528a95fbd145f5654e957
 runc version: v1.1.12-0-g51d5e94
 init version: de40ad0
 Security Options:
  seccomp
   Profile: unconfined
  cgroupns
 Kernel Version: 6.6.31-linuxkit
 Operating System: Docker Desktop
 OSType: linux
 Architecture: aarch64
 CPUs: 8
 Total Memory: 7.657GiB
 Name: docker-desktop
 ID: 41791094-b832-4e43-9b7b-94b72579ceea
 Docker Root Dir: /var/lib/docker
 Debug Mode: false
 HTTP Proxy: http.docker.internal:3128
 HTTPS Proxy: http.docker.internal:3128
 No Proxy: hubproxy.docker.internal
 Labels:
  com.docker.desktop.address=unix:///Users/helen/Library/Containers/com.docker.docker/Data/docker-cli.sock
 Experimental: false
 Insecure Registries:
  hubproxy.docker.internal:5555
  127.0.0.0/8
 Live Restore Enabled: false
```

### Diagnostics ID

4447679C-2A79-4BF9-BA84-45E7E5A84517/20240710053009

**Comment by helenxie-bit** [comment]
Sorry, I forgot to create the corresponding repository on my Docker Hub. However, after creating it and trying to push again, I encountered the following error:
failed to copy: failed to do request: Put "https://registry-1.docker.io/v2/helenxiehz428/test_llm/blobs/uploads/c2e5946c-5b0e-408b-aa77-30fb170200bf?_state=1AowNNVk2cOMXGt9IFQujXKaSHZtBX0B2v3dbB-bO017Ik5hbWUiOiJoZWxlbnhpZWh6NDI4L3Rlc3RfbGxtIiwiVVVJRCI6ImMyZTU5NDZjLTViMGUtNDA4Yi1hYTc3LTMwZmIxNzAyMDBiZiIsIk9mZnNldCI6MCwiU3RhcnRlZEF0IjoiMjAyNC0wNy0xMFQwNTo0MTo1NC4xMTk0OTM1MDJaIn0%3D&digest=sha256%3A9664d484259cfec4dbdb9c37e64afd24eb50def53d7bf7c3d22b8ed20be69a43": write tcp 192.168.65.3:55058->192.168.65.1:3128: use of closed network connection

**Comment by helenxie-bit** [comment]
It failed due to the network problem. I changed to use a new network and it succeeded.

**Comment by geocine** [comment]
> It failed due to the network problem. I changed to use a new network and it succeeded.

how to change to a new network?

**Comment by helenxie-bit** [comment]
@geocine I went a new place and use the network there, and the problem was fixed. I think it needs a more stable network for pushing images to DockerHub.

**Comment by bortvern** [comment]
Just ran into this and it's exactly as @cwadrupldijjit said. On Comcast DOCSIS cable modem connection get this broken pipe error, but on AT&T fiber it works fine.

**Comment by cwadrupldijjit** [comment]
As a follow up to this conversation, I finally have found the solution to the problem for me.  I had a Comcast technician come out to my home today to take a look at it.  It turns out that he found in the logs where it was killing the connections and that was part of their firmware of the modem portion of their internet gateway which is in place to prevent DDOS-style attacks from originating on networks with that gateway or efforts to take up too much bandwidth from anyone else using their service.  It's likely if I swap out the gateway they provided (and we are technically renting) for my own router/modem setup, it likely will get rid of the issue.  We were able to test the setup with a "bridge mode", which essentially turns off most of the modem and router stuff and leaves only things connected to it via ethernet to connect to the internet.  When I did that, it worked flawlessly.

If you are like me (especially nine or ten months in to a 24-month contract with Comcast Business) and can't switch off of your ISP, I suggest getting your own router/modem setup (possibly combined), and it very well may work flawlessly after it.

The technician did mention that if it does happen again, it might be something even deeper in their infrastructure, even less easy or even impossible to circumvent than the just the device I have.  Even if there are other complications, I think it might be tolerable enough for me.
